app/src/main/java/emu/skyline/provider/SkylineDocumentsProvider.kt

```diff
@@ -0,0 +1,307 @@
+            )
+        }
+        return copyDocument(sourceDocumentId, targetParentDocumentId)
+    }
+
+    override fun copyDocument(sourceDocumentId : String, targetParentDocumentId : String?) : String? {
+
+        val parent = getFileForDocId(targetParentDocumentId!!)
+        val oldFile = getFileForDocId(sourceDocumentId)
+        val newFile = File(parent.path, oldFile.name)
+        try {
+
+            var wasNewFileCreated = false
+            if (newFile.createNewFile()) {
+                if (newFile.setWritable(true) && newFile.setReadable(true)) {
+                    wasNewFileCreated = true
+                }
+            }
+            if (!wasNewFileCreated) {
+                throw FileNotFoundException(
+                    "Couldn't copy document " + sourceDocumentId +
+                            ". Could not create new file."
+                )
+            }
+            FileInputStream(oldFile).use { inStream ->
+                FileOutputStream(newFile).use { outStream ->
+                    val buf : ByteArray = ByteArray(4096)
+                    var len : Int
+                    while ((inStream.read(buf).also { len = it }) > 0) {
+                        outStream.write(buf, 0, len)
+                    }
+                }
+            }
+        } catch (e : IOException) {
+            throw FileNotFoundException(
+                ("Failed to copy document: " + sourceDocumentId +
+                        ". " + e.message)
+            )
+        }
+        return getDocIdForFile(newFile).toString()
+    }
+
+    override fun moveDocument(
+        sourceDocumentId : String, sourceParentDocumentId : String?,
+        targetParentDocumentId : String?
+    ) : String? {
+        try {
+
+            val newDocumentId = copyDocument(
+                sourceDocumentId, sourceParentDocumentId!!,
+                targetParentDocumentId
+            )
+
+            removeDocument(sourceDocumentId, sourceParentDocumentId)
+            return newDocumentId
+        } catch (e : FileNotFoundException) {
+            throw FileNotFoundException("Couldn't move document $sourceDocumentId")
+        }
+    }
+
+    private fun includeFile(result : MatrixCursor, documentId : String?, file : File?) : MatrixCursor {
+        var localDocumentId = documentId
+        var localFile = file
+
+        if (localDocumentId == null) {
+            localDocumentId = localFile?.let { getDocIdForFile(it) } as String?
+        } else if (localFile == null) {
+            localFile = getFileForDocId(localDocumentId)
+        }
+
+        var flags = 0
+        if (localFile == null) {
+            throw FileNotFoundException()
+        } else {
+            if (localFile.isDirectory) {
+                if (localFile.isDirectory && localFile.canWrite()) {
+                    flags = flags or DocumentsContract.Document.FLAG_DIR_SUPPORTS_CREATE
+                }
```

Comment:
Any reason to recheck `localFile.isDirectory` here? Additionally, do remove the `{}` braces here.